Serialization and equality comparison of CSS number tokens in custom properties should use the original string

anttijk opened the pull request:
#### 11c57336e700b00fa2a46d465be30274518e0816
<pre>
Serialization and equality comparison of CSS number tokens in custom properties should use the original string
<a href="https://bugs.webkit.org/show_bug.cgi?id=269910">https://bugs.webkit.org/show_bug.cgi?id=269910</a>
<a href="https://rdar.apple.com/123435602">rdar://123435602</a>

Reviewed by Matthieu Dubet.

Custom property value &quot;0.10&quot; should serialize as such and not as &quot;0.1&quot;. Also they shouldn&apos;t compare equal.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/at-container-style-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/custom-property-style-queries-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-custom-props-expected.txt:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::customCSSText const):
* Source/WebCore/css/CSSPropertyRule.cpp:
(WebCore::CSSPropertyRule::initialValue const):
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::serialize const):

Use the custom property serialization mode.

* Source/WebCore/css/CSSVariableData.h:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::customCSSText const):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::CSSParserToken::originalText const):
(WebCore::CSSParserToken::operator== const):

Use string equality for numbers. This is used for custom properties only.

(WebCore::CSSParserToken::serialize const):

Add a separate serialization mode for custom properties. It returns the original string for number values.
It will later also preserve whitespace.

* Source/WebCore/css/parser/CSSParserToken.h:
* Source/WebCore/css/parser/CSSParserTokenRange.cpp:
(WebCore::CSSParserTokenRange::serialize const):
* Source/WebCore/css/parser/CSSParserTokenRange.h:

Serialize the original text.

Canonical link: <a href="https://commits.webkit.org/275196@main">https://commits.webkit.org/275196@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/461c613c68e698624cc3831543153f99146c0c73

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41121 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43499 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43683 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37213 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43428 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17465 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34045 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41695 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17067 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35424 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14684 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14815 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36424 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45001 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37322 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36739 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40487 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15936 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13073 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38863 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17555 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9233 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17607 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17199 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/css/CSSVariableData.h:
```diff
@@ -51,6 +51,8 @@ class CSSVariableData : public RefCounted<CSSVariableData> {
 
     bool operator==(const CSSVariableData& other) const;
 
+    String serialize() const;
```
`const String&` ?

anttijk replied:
No, this must return an owned string.

mdubet commented on Source/WebCore/css/parser/CSSParserToken.h:
```diff
@@ -135,7 +135,11 @@ class CSSParserToken {
+        Normal,
+        CustomProperty // Custom properties don't collapse whitespace and serialize numbers as original strings.
+    };
+    void serialize(StringBuilder&, const CSSParserToken* nextToken = nullptr, SerializationMode = SerializationMode::Normal) const;
```
`SerializationMode` could be a constexpr here it seems?

(albeit we don't care that much about performance for serialisation code anyway)

mdubet commented on Source/WebCore/css/parser/CSSParserToken.cpp:
```diff
@@ -661,19 +665,28 @@ void CSSParserToken::serialize(StringBuilder& builder, const CSSParserToken* nex
-        if (m_numericSign == PlusSign)
-            builder.append('+');
-        builder.append(numericValue());
+        if (mode == SerializationMode::CustomProperty)
```
Same, it looks like it could be a `if constexpr`

anttijk replied:
But it is not a template parameter?

mdubet replied:
It seems it could be. The only time it's called with `SerializationMode::CustomProperty` is statically determined in `CSSVariableData::serialise()`.

anttijk replied:
I don't think templating this is a meaningful optimization.

mdubet replied:
Yes indeed serialisation is really not perf critical enough to the point of avoiding a runtime `if` I guess 👌

anttijk committed:
Serialization and equality comparison of CSS number tokens in custom properties should use the original string
https://bugs.webkit.org/show_bug.cgi?id=269910
rdar://123435602

Reviewed by Matthieu Dubet.

Custom property value "0.10" should serialize as such and not as "0.1". Also they shouldn't compare equal.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/at-container-style-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/container-queries/custom-property-style-queries-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/serialize-custom-props-expected.txt:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::customCSSText const):
* Source/WebCore/css/CSSPropertyRule.cpp:
(WebCore::CSSPropertyRule::initialValue const):
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::serialize const):

Use the custom property serialization mode.

* Source/WebCore/css/CSSVariableData.h:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::customCSSText const):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::CSSParserToken::originalText const):
(WebCore::CSSParserToken::operator== const):

Use string equality for numbers. This is used for custom properties only.

(WebCore::CSSParserToken::serialize const):

Add a separate serialization mode for custom properties. It returns the original string for number values.
It will later also preserve whitespace.

* Source/WebCore/css/parser/CSSParserToken.h:
* Source/WebCore/css/parser/CSSParserTokenRange.cpp:
(WebCore::CSSParserTokenRange::serialize const):
* Source/WebCore/css/parser/CSSParserTokenRange.h:

Serialize the original text.

Canonical link: https://commits.webkit.org/275196@main
